### DISCUSSION TITLE
Problem with UserControl properties

### DISCUSSION THREAD

**Discussion by MichaelSvejcar** [AvaloniaUI/Avalonia#10323] [Q&A]
I am facing problem with custom setting UserControl properties. The UserControl is defined in it's own xaml file and I am using it in the MainWindow view xaml. In it's code behind, I am implementing the INotifyPropertyChanged. Properties in the UserControl code behind are binded to controls in the UserControl view and are implementing INotifyPropertyChanged. However, when using the UserControl in the MainWindow view, binding it's properties to the MainWindowViewModel doesn't get updated the GUI of the UserControl updated, even though the setters of the UserControl properties are getting hit.

This is snippet sample code of the UserControl:

```
public static readonly AvaloniaProperty<bool> IsOpenProperty =
            AvaloniaProperty.Register<PopupControl, bool>(
                nameof(IsOpen),
                defaultValue: false);

        public void OnPropertyChanged(object sender, [CallerMemberName] string? propertyName = null)
        {
            PropertyChanged?.Invoke(sender, new PropertyChangedEventArgs(propertyName));
        }
        public event PropertyChangedEventHandler? PropertyChanged;


        public bool IsOpen
        {
            get => (bool)GetValue(IsOpenProperty);
            set
            {
                SetValue(IsOpenProperty, value);
                OnPropertyChanged(this, nameof(IsOpen));
            }
        }

        public PopupControl()
        {
            InitializeComponent();
            IsOpen = false;
        }
```
What am I missing here?

**Comment by maxkatz6** [comment]
Use OnPropertyChanged virtual method or PropertyChanged event that is part of the Control class. Setter might not be executed. 

Also, avalonia controls implement INotifyPropertyChanged already.

**Comment by timunie** [comment]
Are you sure `UserControl` is what you are looking for? `UserControl` can be seen as a kind of "Page". If you want a reuseable control, you most likely want to base on `ContentControl` or `TemplatedControl`. 

P.S.: I've a draft of a sample here: https://github.com/timunie/Avalonia.Samples/tree/Controls/RatingControl/src/Avalonia.Samples/CustomControls/RatingControlSample

**Reply by MichaelSvejcar** [reply]
I want to use it for Popup template, that will implement style of it's border and buttons, so that every time I use it and add different buttons and their amount, there won't be no need to implement the style for each one individually. I also want to use UserControl for window template in a similar way. Is that good choice for this scenario?

**Reply by timunie** [reply]
I think in that case I'd base on `ContentControl` instead.
